src/tribol/physics/Mortar.cpp

```diff
@@ -631,6 +641,565 @@ void ComputeSingleMortarJacobian( SurfaceContactElem& elem )
+          LinIsoQuadShapeFunc( xi2[0], xi2[1], l, phiB );
+          mortar_mat2[k * size2 + l] += phiA * phiB * quad_wt;
+        }
+      }
+
+      // Updated numbering version begin...
+      // RealT phi1[4];
+      // // NOTE: this limits this routine to quads
+      // LinIsoQuadShapeFunc( xi1, phi1 );
+      // for ( int k{ 0 }; k < size1; ++k ) {
+      //   for ( int l{ 0 }; l < size1; ++l ) {
+      //     mortar_mat1[k * size1 + l] += phi1[k] * phi1[l] * quad_wt;
+      //   }
+      // }
+
+      // // 5. Evaluate mortar matrix (nonmortar/mortar contribs)
+      // RealT phi2[4];
+      // // NOTE: this limits this routine to quads
+      // LinIsoQuadShapeFunc( xi2, phi2 );
+      // for ( int k{ 0 }; k < size1; ++k ) {
+      //   for ( int l{ 0 }; l < size2; ++l ) {
+      //     mortar_mat2[k * size2 + l] += phi1[k] * phi2[l] * quad_wt;
+      //   }
+      // }
+      // Updated numbering version end...
+    }
+  }
+
+  // compute gaps
+  for ( int i{ 0 }; i < size1; ++i ) {
+    g1[i] = 0.0;
+    RealT gap_v[3] = { 0.0, 0.0, 0.0 };
+    for ( int j{ 0 }; j < size1; ++j ) {
+      for ( int d{ 0 }; d < 3; ++d ) {
+        gap_v[d] -= mortar_mat1[i * size1 + j] * x1[d * size1 + j];
+      }
+    }
+    for ( int j{ 0 }; j < size2; ++j ) {
+      for ( int d{ 0 }; d < 3; ++d ) {
+        gap_v[d] += mortar_mat2[i * size2 + j] * x2[d * size2 + j];
+      }
+    }
+    for ( int d{ 0 }; d < 3; ++d ) {
+      g1[i] += n1[d * size1 + i] * gap_v[d];
+    }
+  }
+
+  // compute nonmortar force contributions
+  for ( int i{ 0 }; i < size1; ++i ) {
+    for ( int d{ 0 }; d < 3; ++d ) {
+      f1[d * size1 + i] = 0.0;
+    }
+    for ( int j{ 0 }; j < size1; ++j ) {
+      for ( int d{ 0 }; d < 3; ++d ) {
+        f1[d * size1 + i] -= p1[j] * n1[d * size1 + i] * mortar_mat1[j * size1 + i];
+      }
+    }
+  }
+
+  // compute mortar force contributions
+  for ( int i{ 0 }; i < size2; ++i ) {
+    for ( int d{ 0 }; d < 3; ++d ) {
+      f2[d * size2 + i] = 0.0;
+    }
+    for ( int j{ 0 }; j < size1; ++j ) {
+      for ( int d{ 0 }; d < 3; ++d ) {
+        f2[d * size2 + i] += p1[j] * n1[d * size1 + i] * mortar_mat2[j * size2 + i];
+      }
+    }
+  }
+}
+
+//------------------------------------------------------------------------------
+void ComputeMortarJacobianEnzyme( const RealT* x1, const RealT* n1, const RealT* p1, RealT* f1, RealT* df1dx1,
+                                  RealT* df1dx2, RealT* df1dn1, RealT* df1dp1, RealT* g1, RealT* dg1dx1, RealT* dg1dx2,
+                                  RealT* dg1dn1, int size1, const RealT* x2, RealT* f2, RealT* df2dx1, RealT* df2dx2,
+                                  RealT* df2dn1, RealT* df2dp1, int size2 )
+{
+  RealT x1_dot[12] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };
+  for ( int i{ 0 }; i < size1 * 3; ++i ) {
+    x1_dot[i] = 1.0;
+    // clang-format off
+      __enzyme_fwddiff<void>((void*)ComputeMortarForceEnzyme,
```

Comment:
I see now that the jacobian routine has calls to the mortar force routine embeded within the enzyme call. Please clarify in the ApplyNormalRoutine() just to orient someone looking at that for the first time.

Reply:
Can you clarify what you mean?